rfcs/kubernetes-oci/README.md

```diff
@@ -0,0 +1,251 @@
+how OCI artifacts are pulled from container registries and how the artifact's authenticity can be verified.
+Flux should be able to work with any type of artifact even if it's not created with the Flux CLI.
+
+## Motivation
+
+Given that OCI registries are evolving into a generic artifact storage solution,
+we should extend Flux to allow fetching Kubernetes manifests and related configs
+from container registries similar to how Flux works with Git and Bucket storage.
+
+With OCI support, Flux users can automate artifact updates to Git in the same way
+they do today for container images.
+
+### Goals
+
+- Add support to the Flux CLI for packaging Kubernetes manifests and related configs into OCI artifacts.
+- Add support to Flux source-controller for fetching configs stored as OCI artifacts.
+- Make it easy for users to switch from Git repositories and Buckets to OCI repositories.
+
+### Non-Goals
+
+- Introduce a new OCI media type for artifacts containing Kubernetes manifests.
+
+## Proposal
+
+### Push artifacts
+
+Flux users should be able to package a local directory containing Kubernetes configs into a tarball
+and push the archive to a container registry as an OCI artifact.
+
+```sh
+flux push artifact docker.io/org/app-config:v1.0.0 -f ./deploy
+```
+
+Flux CLI with produce artifacts of type `application/vnd.oci.image.config.v1+json`.
+The directory pointed to by `-f` is archived and compressed in the `tar+gzip` format
+and the layer media type is set to `application/vnd.oci.image.layer.v1.tar+gzip`.
+
+> A proof-of-concept CLI implementation for distributing Kubernetes configs as OCI artifacts
+> is available at [kustomizer.dev](https://github.com/stefanprodan/kustomizer).
+
+### Pull artifacts
+
+Flux users should be able to define a source for pulling manifests inside the cluster from an OCI repository.
+
+```yaml
+apiVersion: source.toolkit.fluxcd.io/v1beta2
+kind: OCIRepository
+metadata:
+  name: app-config
+  namespace: flux-system
+spec:
+  interval: 10m
+  url: docker.io/org/app-config
+  ref:
+    tag: v1.0.0
+```
+
+An `OCIRepository` can refer to an artifact by tag, digest or semver range:
+
+```yaml
+spec:
+  ref:
+    # one of
+    tag: "latest"
+    digest: "sha256:45b23dee08af5e43a7fea6c4cf9c25ccf269ee113168c19722f87876677c5cb2"
+    semver: "6.0.x"
+```
+
+To verify the authenticity of an artifact, the Sigstore cosign public key can be supplied with:
+
+```yaml
+spec:
+  verify:
+    provider: cosign
+    secretRef:
+      name: cosign-key
+```
+
+### Pull artifacts from private repositories
+
+For authentication purposes, Flux users can choose between supplying static credentials with Kubernetes secrets
+and cloud-based OIDC using an IAM role binding to the source-controller Kubernetes service account.
+
+For private repositories hosted on DockerHub, GitHub, Quay, self-hosted Docker Registry and others,
+the credentials can be supplied with:
+
+```yaml
+spec:
+  secretRef:
```

Comment:
I think a `ServiceAccount` resource can be referenced here, in addition to `secretRef` to access registry credentials. Can the spec look something like this?
```yaml
spec:
  auth:  
    # one of
    secretRef:
      name: regcred
    serviceAccountName: reg
  certSecretRef: regcert
```

Reply:
Maybe in the next iteration of the source API `v1beta3`, right now all APIs have `secretRef` in `spec`. If we want to get this into Flux before GA, then we need to conform to the current API standard.

Reply:
Sure, we can iterate on this in `v1beta3` and keep it as a suggestion.

Reply:
@stefanprodan does it make sense to add `spec.serviceAccountName` in this API proposal?

Reply:
We need to add it to ImageRepository first, could you contribute that to image reflector?

Reply:
Yes @stefanprodan I can add `spec.serviceAccountName` in `ImageRepository` v1beta1 API and implement it in `image-reflector-controller`.  I will put up a PR for the `image-reflector-controller` in the next few days. That said, if you accept the `spec.serviceAccountName` for `OCIRepository` contingent on the PR for the `ImageRepository`, that will be great.

Reply:
Yeap I think `serviceAccountName` is a great addition to both image and source APIs, for consistency sake, we'll need this into image v1beta1 API. Thanks for your help.

Reply:
@stefanprodan logged an enhancement to include `ServiceAccountName` for `ImageRepository` v1beta1 API in image reflector controller https://github.com/fluxcd/image-reflector-controller/issues/251. I will submit the PR for [#251](https://github.com/fluxcd/image-reflector-controller/issues/251) by next week.

Reply:
I've added `serviceAccountName` to the proposal, thanks @rashedkvm for the IRC PR.